### ISSUE TITLE
Crash when joining or leaving group call due to invalid source

### ISSUE THREAD

**Issue by anthonybilinski** [qTox/qTox#5843]
##### Brief Description

OS: Ubuntu 19.04
qTox version: v1.16.3+
Commit hash:  31f1c746b65015ca2d44b0236984fa244459e3fa
toxcore:  v0.2.10
Qt:  5.12.2
Hardware:  Desktop

Reproducible: Sometimes

##### Steps to reproduce

1. Load persistent group.
2. Join existing group call that has a few other members in it, already chatting.

or

1. Be in group audio call
2. Leave group audio call

##### Observed Behavior
Crash on call join:
```
[05:20:30.477 UTC] audio/backend/openal.cpp:583 : Debug: Closing audio input
[05:20:30.873 UTC] core/coreav.cpp:562 : Debug: "Joining group call 4"
[05:20:30.873 UTC] audio/backend/openal.cpp:397 : Debug: Opening audio input "Yeti Stereo Microphone Analog Stereo"
[New Thread 0x7fffd57fa700 (LWP 28159)]
[Thread 0x7fffb261c700 (LWP 28058) exited]
[Thread 0x7fffd57fa700 (LWP 28056) exited]

Thread 11 "qTox Core" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffd4e63700 (LWP 28014)]
ToxGroupCall::playAudioBuffer (this=0x55555e979450, peer=..., data=0x7fffb8013a70, samples=960, channels=2, sampleRate=48000)
    at /home/abilinski/builds/qTox/src/core/toxcall.cpp:330
330             source->second->playAudioBuffer(data, samples, channels, sampleRate);
```
Crash on call end:
```
[06:55:07.976 UTC] core/coreav.cpp:583 : Debug: "Leaving group call 5"
[06:55:07.977 UTC] audio/backend/openal.cpp:306 : Debug: Audio source 3 deleted. Sources active: 2
[06:55:07.977 UTC] audio/backend/openal.cpp:306 : Debug: Audio source 1 deleted. Sources active: 1
[06:55:07.977 UTC] audio/backend/openal.cpp:306 : Debug: Audio source 4 deleted. Sources active: 0
[Thread 0x7fffb17dc700 (LWP 28553) exited]
[06:55:07.980 UTC] audio/backend/openal.cpp:609 : Debug: Closing audio output
[06:55:07.996 UTC] audio/backend/openal.cpp:359 : Debug: Unsubscribed from audio input device [ 0 subscriptions left ]
[06:55:07.996 UTC] audio/backend/openal.cpp:583 : Debug: Closing audio input
[Thread 0x7fffb2204700 (LWP 28552) exited]
[Thread 0x7fffb2a05700 (LWP 28551) exited]
[Thread 0x7fffd57fa700 (LWP 28550) exited]

Thread 11 "qTox Core" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffd4e63700 (LWP 28516)]
ToxGroupCall::playAudioBuffer (this=0x55555e7b7df0, peer=..., data=0x7fffb8034ca0, samples=960, channels=2, sampleRate=48000)
    at /home/abilinski/builds/qTox/src/core/toxcall.cpp:330
330             source->second->playAudioBuffer(data, samples, channels, sampleRate);
(gdb) bt
#0  0x00005555558b94f5 in ToxGroupCall::playAudioBuffer(ToxPk const&, short const*, int, unsigned int, int)
    (this=0x55555e7b7df0, peer=..., data=0x7fffb8034ca0, samples=960, channels=2, sampleRate=48000) at /home/abilinski/builds/qTox/src/core/toxcall.cpp:330
#1  0x000055555589ccaa in CoreAV::groupCallCallback(void*, unsigned int, unsigned int, short const*, unsigned int, unsigned char, unsigned int, void*)
    (tox=0x5555562a5880, group=5, peer=2, data=0x7fffb8034ca0, samples=960, channels=2 '\002', sample_rate=48000, core=0x55555614a8a0)
    at /home/abilinski/builds/qTox/src/core/coreav.cpp:521
#2  0x00007ffff627fafb in decode_audio_packet (group_av=0x55555fb18f60, peer_av=0x7fffb801bbd0, groupnumber=5, friendgroupnumber=2)
    at /home/abilinski/builds/toxcore/toxav/groupav.c:394
```
##### Additional Info
I've only seen each one once. Likely some kind of race condition.

**Comment by anthonybilinski** [comment]
I reprod this in a couple seconds by just clicking the call button quickly in a group. This time the backtrace looked like:
```
#0  0x00007ffff426ea99 in qstrcmp(QByteArray const&, QByteArray const&) ()                                                                                                                                  
    at /usr/lib/x86_64-linux-gnu/libQt5Core.so.5                                                                                                                                                            
#1  0x00005555558c6d6b in operator<(QByteArray const&, QByteArray const&)                                                                                                                                   
    (a1=..., a2=...)                                                                                                                                                                                        
    at /usr/include/x86_64-linux-gnu/qt5/QtCore/qbytearray.h:633                                                                                                                                            
#2  0x00005555558c6ba4 in ContactId::operator<(ContactId const&) const                                                                                                                                      
    (this=0x7fffc6ff9ea0, other=...)                                                                                                                                                                        
    at /home/abilinski/builds/qTox/src/core/contactid.cpp:70                                                                                                                                                
#3  0x00005555558be230 in std::less<ToxPk>::operator()(ToxPk const&, ToxPk const&) const (this=0x5555577402f0, __x=..., __y=...)                                                                            
--Type <RET> for more, q to quit, c to continue without paging--                                                                                                                                            
   l_function.h:386                                                                                                                                                                                         
#4  0x00005555558c0155 in std::_Rb_tree<ToxPk, std::pair<ToxPk const, QMetaObject::Connection>, std::_Select1st<std::pair<ToxPk const, QMetaObject::Connection> >, std::less<ToxPk>, std::allocator<std::pa$r<ToxPk const, QMetaObject::Connection> > >::_M_get_insert_unique_pos(ToxPk const&) (this=0x5555577402f0, __k=...) at /usr/include/c++/8/bits/stl_tree.h:2054                                              
#5  0x00005555558be624 in std::_Rb_tree<ToxPk, std::pair<ToxPk const, QMetaObject::Connection>, std::_Select1st<std::pair<ToxPk const, QMetaObject::Connection> >, std::less<ToxPk>, std::allocator<std::pa$r<ToxPk const, QMetaObject::Connection> > >::_M_insert_unique<std::pair<ToxPk const, QMetaObject::Connection> >(std::pair<ToxPk const, QMetaObject::Connection>&&) (this=0x5555577402f0, __v=...)          
    at /usr/include/c++/8/bits/stl_tree.h:2106
#6  0x00005555558bd4b3 in std::map<ToxPk, QMetaObject::Connection, std::less<ToxPk>, std::allocator<std::pair<ToxPk const, QMetaObject::Connection> > >::insert(std::pair<ToxPk const, QMetaObject::Connect$on>&&) (this=0x5555577402f0, __x=...) at /usr/include/c++/8/bits/stl_map.h:809
#7  0x00005555558b92c8 in ToxGroupCall::addPeer(ToxPk) (this=0x555557740270, peerId=...) at /home/abilinski/builds/qTox/src/core/toxcall.cpp:303                                                           
#8  0x00005555558b94fa in ToxGroupCall::playAudioBuffer(ToxPk const&, short const*, int, unsigned int, int) (this=0x555557740270, peer=..., data=0x7fffb8019d00, samples=960, channels=2, sampleRate=48000)
    at /home/abilinski/builds/qTox/src/core/toxcall.cpp:326
#9  0x000055555589cca4 in CoreAV::groupCallCallback(void*, unsigned int, unsigned int, short const*, unsigned int, unsigned char, unsigned int, void*)                                                     
    (tox=0x555556087250, group=0, peer=1, data=0x7fffb8019d00, samples=960, channels=2 '\002', sample_rate=48000, core=0x555556174850) at /home/abilinski/builds/qTox/src/core/coreav.cpp:521              
```

I think what's happening here is that on the CoreAv thread, we have:
```
CoreAV::groupCallCallback
	groupCalls[group]
ToxGroupCall::playAudioBuffer
```
and on the GUI thread, we have:
```
GroupChatForm::leaveGroupCall
CoreAV::leaveGroupCall
	groupCalls.erase
ToxGroupCall::~ToxGroupCall
ToxGroupCall::clearPeers
```

so `CoreAV::groupCallCallback` is grabbing a reference to a ToxCall that is then deleted from under it.

Looking at CoreAv's functions that touch the `groupCalls` map, there are:

| Function  | Thread called from |
| ------------- | ------------- |
| groupCallCallback  | CoreAv  |
| invalidateGroupCallPeerSource  | Widget  |
| joinGroupCall  | GUI, Widget, QApplication::aboutToQuit  |
| leaveGroupCall  | Content Cell  |
| sendGroupCallAudio  | CoreAv  |
| muteCallInput  | GUI  |
| muteCallOutput  | GUI  |
| isGroupCallInputMuted  | GUI  |
| isGroupCallOutputMuted  | GUI  |

sooo I think we need turn what are a bunch of singleton calls into CoreAv to instead be signal/slots, so that all accesses to class (or in this case file static?) state can be on the CoreAv thread.

**Comment by anthonybilinski** [comment]
I hit this a few more times in a single group call, it seems more common to repro in a group with more members, (I had 4 this time compared to 3 before). Combined with https://github.com/TokTok/c-toxcore/issues/961#issuecomment-533895287, this is pretty painful because when you try to leave a call, you often crash, causing you to become mute in that group for good. When you create a new group, everyone has to join the call, and there's a good chance one of them will crash again.

**Comment by sphaerophoria** [comment]
Lucked out and got a backtrace today

```
[03:08:39.894 UTC] audio/backend/openal.cpp:278 : Debug: Audio source 1 created. Sources active: 1

Thread 13 "qTox Core" received signal SIGSEGV, Segmentation fault.
[Switching to LWP 171614]
0x00007ffff36a1a43 in ?? ()
(gdb) bt
#0  0x00007ffff36a1a43 in ?? ()
#1  0x0000555555826572 in std::_Rb_tree_iterator<std::pair<ToxPk const, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> > > >::operator-- (this=0x7fffbaba8600) at /usr/include/c++/9.2.0/bits/stl_tree.h:302
#2  0x00005555558251a7 in std::_Rb_tree<ToxPk, std::pair<ToxPk const, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> > >, std::_Select1st<std::pair<ToxPk const, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> > > >, std::less<ToxPk>, std::allocator<std::pair<ToxPk const, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> > > > >::_M_get_insert_unique_pos (this=0x5555567c62c0, __k=...) at /usr/include/c++/9.2.0/bits/stl_tree.h:2104
#3  0x000055555582391e in std::_Rb_tree<ToxPk, std::pair<ToxPk const, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> > >, std::_Select1st<std::pair<ToxPk const, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> > > >, std::less<ToxPk>, std::allocator<std::pair<ToxPk const, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> > > > >::_M_emplace_unique<ToxPk&, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> > > (this=0x5555567c62c0, __args#0=..., __args#1=...) at /usr/include/c++/9.2.0/bits/stl_tree.h:2413
#4  0x0000555555822541 in std::map<ToxPk, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> >, std::less<ToxPk>, std::allocator<std::pair<ToxPk const, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> > > > >::emplace<ToxPk&, std::unique_ptr<IAudioSink, std::default_delete<IAudioSink> > > (this=0x5555567c62c0, __args#0=..., __args#1=...) at /usr/include/c++/9.2.0/bits/stl_map.h:575
#5  0x000055555581e3ad in ToxGroupCall::addPeer (this=0x5555567c6270, peerId=...) at ../src/core/toxcall.cpp:302
#6  0x000055555581e622 in ToxGroupCall::playAudioBuffer (this=0x5555567c6270, peer=..., data=0x7fffb0007130, samples=960, channels=2, sampleRate=48000) at ../src/core/toxcall.cpp:326
#7  0x00005555557fc65e in CoreAV::groupCallCallback (tox=0x5555562b8000, group=0, peer=1, data=0x7fffb0007130, samples=960, channels=2 '\002', sample_rate=48000, core=0x5555564ed090) at ../src/core/coreav.cpp:503
```

I haven't managed to find a consistent repro case yet

**Comment by sphaerophoria** [comment]
I'm pretty sure this is just a race around CoreAV not protecting itself against multiple threads. The gui thread will call "leaveGroupCall" which erases the group call from the groupCalls member. This could be happening while we're in our groupCallCallback which would end up with us getting an invalid iterator + pointer to our ToxGroupCall instance.

I don't have good proof for this really since the repro is hard, but I can prove that the calls happen on different threads with no protection which I think should be enough to justify guarding against the race.

**Comment by sphaerophoria** [comment]
Whoops my apologies @anthonybilinski. I thought this was closed with a fix and re-opened cause it's still happening. Looking back that's not the case and I've just duplicated your comments.

**Comment by sudden6** [comment]
This should be fixed by #5926 can someone confirm?

**Comment by anthonybilinski** [comment]
Confirmed this is fixed by #5926. My testing was:
1) Create an audio group with 8 qTox profiles
2) Group call with all 8, sending audio and playing audio to make sure we're exercising the path
3) On one profile, click join/leave call as fast as possible.

Testing on `v1.17-dev`, which doesn't have this fix, reliably the profile that's joining and leaving the call segfaults in about 3s with the same ToxGroupCall::playAudioBuffer -> ToxGroupCall::addPeer backtrace. I reproduced this fast crash about 10 times. Updating that profile's qTox to master which does include #5926, clicking as fast as I could on join/leave call for about 2 minutes, I saw no crashes.

I'll cherry-pick this fix over to `v1.17-dev` since this issue is part of our release milestone.
